crates/valence_command/src/scopes.rs

```diff
@@ -0,0 +1,281 @@
+//! Scope graph for the Valence Command system.
+//!
+//! ## Breakdown
+//! Each scope is a node in the graph. a path from one node to another indicates
+//! that the first scope implies the second. A colon in the scope name indicates
+//! a sub-scope. you can use this to create a hierarchy of scopes. for example,
+//! the scope "valence:command" implies "valence:command:tp". this means that if
+//! a player has the "valence:command" scope, they can use the "tp" command.
+//!
+//! You may also link scopes together in the registry. this is useful for admin
+//! scope umbrellas. for example, if the scope "valence:admin" is linked to
+//! "valence:command", It means that if a player has the "valence:admin" scope,
+//! they can use all commands under the command scope.
+//!
+//! # Example
+//! ```
+//! use valence_command::scopes::CommandScopeRegistry;
+//!
+//! let mut registry = CommandScopeRegistry::new();
+//!
+//! // add a scope to the registry
+//! registry.add_scope("valence:command:teleport");
+//!
+//! // we added 4 scopes to the registry. "valence", "valence:command", "valence:command:teleport",
+//! // and the root scope.
+//! assert_eq!(registry.scope_count(), 4);
+//!
+//! registry.add_scope("valence:admin");
+//!
+//! // add a scope to the registry with a link to another scope
+//! registry.link("valence:admin", "valence:command:teleport");
+//!
+//! // the "valence:admin" scope implies the "valence:command:teleport" scope
+//! assert_eq!(
+//!     registry.grants("valence:admin", "valence:command:teleport"),
+//!     true
+//! );
+//! ```
+
+use std::collections::HashMap;
+use std::fmt::{Debug, Formatter};
+
+use bevy_ecs::prelude::Component;
+use bevy_ecs::system::Resource;
+use petgraph::dot;
+use petgraph::dot::Dot;
+use petgraph::prelude::*;
+
+/// Command scope Component for players. this is a list of scopes that a player
+/// has. if a player has a scope, they can use any command that requires
+/// that scope.
+#[derive(Debug, Clone, PartialEq, Eq, Hash, Ord, PartialOrd, Component, Default)]
+pub struct CommandScopes(pub Vec<String>);
```

Comment:
This should derive `Deref`, `DerefMut` since that has become the established convention.

Reply:
if this is common maybe a derive macro would be useful for this.

Reply:
There is. You should use the macro from `derive_more` or `bevy_derive`.